File: backend.scala/app/com/skewwhiffy/auraltester/notes/AbsoluteNote.scala
package com.skewwhiffy.auraltester.notes

import com.skewwhiffy.auraltester.scales.Key

import scala.util.chaining.scalaUtilChainingOps

object AbsoluteNote {
  lazy val middleC: AbsoluteNote = AbsoluteNote(Note.c, Octave.default)
}

case class AbsoluteNote(
  note: Note,
  octave: Octave,
  lyric: Option[String] = None
) {
  def apply(interval: DirectedInterval): AbsoluteNote = {
    interval.direction match {
      case IntervalDirection.`up` => add(interval.interval)
      case IntervalDirection.`down` => subtract(interval.interval)
    }
  }

  def withLyric(lyric: String): AbsoluteNote = copy(lyric = Some(lyric))

  def +(interval: Interval): AbsoluteNote = add(interval)

  private def add(interval: Interval): AbsoluteNote = {
    val defaultNote: AbsoluteNote = interval.degree match {
      case 1 => this
      case 2 => upMajorSecond
      case 3 => upMajorSecond.upMajorSecond
      case 4 => add(Interval.major(3)).upMinorSecond
      case 5 => add(Interval.perfect(4)).upMajorSecond
      case 6 => add(Interval.perfect(5)).upMajorSecond
      case 7 => add(Interval.major(6)).upMajorSecond
      case 8 => AbsoluteNote(note, octave.up)
      case _ => throw new IllegalArgumentException()
    }

    interval.deviation match {
      case it if it < 0 => List.range(0, -it).foldRight(defaultNote)((_, note) => note.flat)
      case it if it > 0 => List.range(0, it).foldRight(defaultNote)((_, note) => note.sharp)
      case 0 => defaultNote
    }
  }

  def -(interval: Interval): AbsoluteNote = subtract(interval)

  private def subtract(interval: Interval): AbsoluteNote = {
    val defaultNote: AbsoluteNote = interval.degree match {
      case 1 => this
      case 2 => downMajorSecond
      case 3 => downMajorSecond.downMajorSecond
      case 4 => subtract(Interval.major(3)).downMinorSecond
      case 5 => subtract(Interval.perfect(4)).downMajorSecond
      case 6 => subtract(Interval.perfect(5)).downMajorSecond
      case 7 => subtract(Interval.major(6)).downMajorSecond
      case 8 => AbsoluteNote(note, octave.down)
      case _ => throw new IllegalArgumentException()
    }

    interval.deviation match {
      case it if it < 0 => List.range(0, -it).foldRight(defaultNote)((_, note) => note.sharp)
      case it if it > 0 => List.range(0, it).foldRight(defaultNote)((_, note) => note.flat)
      case 0 => defaultNote
    }
  }

  def abc(key: Key): String = key.abc(this)

  def wordAbc: String = if (lyric.getOrElse("").isEmpty) "*" else lyric.get

  def sharp: AbsoluteNote = AbsoluteNote(note.sharp, octave)

  def flat: AbsoluteNote = AbsoluteNote(note.flat, octave)

  def <=(other: AbsoluteNote): Boolean = this < other || this == other

  def >=(other: AbsoluteNote): Boolean = this > other || this == other

  def <(other: AbsoluteNote): Boolean = this.octave < other.octave ||
    (this.octave == other.octave && this.note < other.note)

  def >(other: AbsoluteNote): Boolean = this.octave > other.octave ||
    (this.octave == other.octave && this.note > other.note)

  override def toString: String = abc(Key.cMajor)

  private lazy val upMinorSecond = upMajorSecond
    .pipe(it => AbsoluteNote(it.note.flat, it.octave))

  private lazy val downMinorSecond = downMajorSecond
    .pipe(it => AbsoluteNote(it.note.sharp, it.octave))

  private lazy val upMajorSecond = (if ("B" == note.noteName) octave.up else octave)
    .pipe(it => AbsoluteNote(note.upMajorSecond, it))

  private lazy val downMajorSecond = (if ("C" == note.noteName) octave.down else octave)
    .pipe(it => AbsoluteNote(note.downMajorSecond, it))
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/AbsoluteNote.kt
package com.skewwhiffy.notation.model

data class AbsoluteNote(val note: Note, val octave: Octave, val lyric: String? = null) {
  companion object {
    val middleC: AbsoluteNote = AbsoluteNote(Note.c, Octave.default)
  }

  /*

  def apply(interval: DirectedInterval): AbsoluteNote = {
    interval.direction match {
      case IntervalDirection.`up` => add(interval.interval)
      case IntervalDirection.`down` => subtract(interval.interval)
    }
  }

*/
  fun withLyric(lyric: String): AbsoluteNote = copy(lyric = lyric)

  operator fun plus(interval: Interval): AbsoluteNote {
    val defaultNote: AbsoluteNote = when (interval.degree) {
      1 -> this
      2 -> upMajorSecond
      3 -> upMajorSecond.upMajorSecond
      4 -> (this + Interval.major(3)).upMinorSecond
      5 -> (this + Interval.perfect(4)).upMajorSecond
      6 -> (this + Interval.perfect(5)).upMajorSecond
      7 -> (this + Interval.major(6)).upMajorSecond
      8 -> AbsoluteNote(note, octave.up)
      else -> throw IllegalArgumentException()
    }

    return interval.deviation.let {
      when {
        it < 0 -> (0..-it).fold(defaultNote) { note, _ -> note.flat }
        it > 0 -> (1..it).fold(defaultNote) { note, _ -> note.sharp }
        else -> defaultNote
      }
    }
  }

  operator fun minus(interval: Interval): AbsoluteNote {
    val defaultNote: AbsoluteNote = when (interval.degree) {
      1 -> this
      2 -> downMajorSecond
      3 -> downMajorSecond.downMajorSecond
      4 -> (this - Interval.major(3)).downMinorSecond
      5 -> (this - Interval.perfect(4)).downMajorSecond
      6 -> (this - Interval.perfect(5)).downMajorSecond
      7 -> (this - Interval.major(6)).downMajorSecond
      8 -> AbsoluteNote(note, octave.down)
      else -> throw IllegalArgumentException()
    }

    return interval.deviation.let {
      when {
        it < 0 -> (1..-it).fold(defaultNote) { note, _ -> note.sharp }
        it > 0 -> (1..it).fold(defaultNote) { note, _ -> note.flat }
        else -> defaultNote
      }
    }
  }

  fun abc(key: Key): String = key.abc(this)

  val wordAbc: String = (lyric ?: "").ifEmpty { "*" }

  private val sharp: AbsoluteNote get() = AbsoluteNote(note.sharp, octave)

  private val flat: AbsoluteNote get() = AbsoluteNote(note.flat, octave)

  /*
  def <=(other: AbsoluteNote): Boolean = this < other || this == other

  def >=(other: AbsoluteNote): Boolean = this > other || this == other

  def <(other: AbsoluteNote): Boolean = this.octave < other.octave ||
    (this.octave == other.octave && this.note < other.note)

  def >(other: AbsoluteNote): Boolean = this.octave > other.octave ||
    (this.octave == other.octave && this.note > other.note)

  override def toString: String = abc(Key.cMajor)
  */
  private val upMajorSecond
    get() = (if ("B" == note.noteName) octave.up else octave)
      .let { AbsoluteNote(note.upMajorSecond, it) }

  private val upMinorSecond
    get() = upMajorSecond.let { AbsoluteNote(it.note.flat, it.octave) }

  private val downMajorSecond
    get() = (if ("C" == note.noteName) octave.down else octave)
      .let { AbsoluteNote(note.downMajorSecond, it) }

  private val downMinorSecond
    get() = downMajorSecond.let { AbsoluteNote(it.note.sharp, it.octave) }

}